docs/architecture/architecture.md

```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+arguments, so it needs the help of `CRI-O` to be able to distinguish between a pod
+creation from a container one.
+
+### OCI annotations
+
+In order for the Clear Containers runtime (or any virtual machine  based OCI compatible
+runtime) to be able to understand if it needs to create a full virtual machine or if it
+has to create a new container inside an existing pod's virtual machine, `CRI-O` adds
+specific annotations to the OCI configuration file (`config.json`) which is passed to
+the OCI compatible runtime.
+
+Before calling its runtime, `CRI-O` will always add a `io.kubernetes.cri-o.ContainerType`
+annotation to the `config.json` `OCI` configuration file it produces from the kubelet `CRI`
+request. The `io.kubernetes.cri-o.ContainerType` annotation can either be set to `sandbox`
+or `container`. Clear Containers will then use this annotation to decide if it needs to
+respectively create a virtual machine or a container inside a virtual machine associated
+with a Kubernetes pod:
+
+```Go
+	containerType, err := ociSpec.ContainerType()
+	if err != nil {
+		return err
+	}
+
+	switch containerType {
+	case vc.PodSandbox:
+		process, err = createPod(ociSpec, runtimeConfig, containerID, bundlePath, console, disableOutput)
+		if err != nil {
+			return err
+		}
+	case vc.PodContainer:
+		process, err = createContainer(ociSpec, containerID, bundlePath, console, disableOutput)
+		if err != nil {
+			return err
+		}
+	}
+
+```
+
+### Generalization
+
+Clear Containers 2.1, implemented the initial Kubernetes / `CRI-O` support by choosing
+to handle pods as a special case for containers. By default the runtime would only be aware
+of containers and only when receiving CRI-O specific annotations would it follow specific
+code paths to handle pods and intra pod containers. This approach showed that container
+orchestration support was an afterthought with 2.1, and that supporting the single container
+Docker use case was the main focus.
+
+Starting from the 3.0 release, Clear Containers took a different, almost opposite approach.
+The Clear Containers 3.0 runtime is based on the `virtcontainers` package whose API is
+inspired by the Kubernetes `CRI` one. `virtcontainers` execution units are pods and it does
+not create or manage any container outside of an existing pod context, and thus so does the
+Clear Container 3.0 runtime.
+To create a Docker container it will first create a pod which will represent the container
+hardware virtualized context and then start an actual container inside that virtual machine.
+Both the pod and its container have the same identifier and name.
+
+### Mixing VM based and namespace based runtimes
+
+One interesting evolution of the `CRI-O` support for the Clear Containers runtime is the ability
+to run virtual machine based pods alongside namespace ones. With `CRI-O` and Clear Containers,
+one can introduce the concept of workload trust inside a Kubernetes cluster.
```

Comment:
I'm not sure if "workload trust" should in fact have double-quotes around it as it's introducing a term, so will leave to @iphutch to check ;)